### ISSUE TITLE
-深度学习基础模型介绍.

### ISSUE THREAD

**Issue by rattlesnakey** [rattlesnakey/rattlesnakey.github.io#9]
https://rattlesnakey.github.io/%E6%B7%B1%E5%BA%A6%E5%AD%A6%E4%B9%A0%E5%9F%BA%E7%A1%80%E6%A8%A1%E5%9E%8B%E4%BB%8B%E7%BB%8D/ 

 
  由于课程要求，刚好趁这个机会总结一下常用的深度学习基础模型和激活函数

Model
MLP(Muti-Layer Perception)

  
    MLP包括包括三层：输入层、隐藏层和输出层，MLP神经网络不同层之间是全连接的(全连接的意思就是：上一层的任何一个神经元与下一层的所有神经元都有连接)
  

MLP中包括权重、偏置和激活函数，这三个是构成MLP的基础

  权重(weight): 下一层的每个神经元和上一层的每个神经元之间都有权重值，当前神经元的值是由上一层所有神经元和其对应的权重的线性加权求和而得到的
  偏置(bias): 权重加权求和完之后再加上bias来得到当前神经元的值, 权重+偏置构成线性函数，即\(f(x) = w^Tx + b\)
  激活函数(activation function): 起非线性映射的作用，其可将神经元的输出幅度限制在一定范围内，一般限制在（-1~1）或（0~1）之间。
    MLP代码(Pytorch)
    # 建立一个四层感知机网络
class MLP(torch.nn.Module):  
def __init__(self, in_dim, h1_dim, h2_dim, out_dim):
  super(MLP, self).__init__()  
  self.fc1 = torch.nn.Linear(in_dim, h1_dim) # 第一个隐含层
  self.fc2 = torch.nn.Linear(h1_dim, h2_dim) # 第二个隐含层
  self.fc3 = torch.nn.Linear(h2_dim, out_dim)  # 输出层
 
def forward(self, din):
  # 前向传播， 输入值：din, 返回值 dout
  dout = F.relu(self.fc1(din))  
  dout = F.relu(self.fc2(dout))
  dout = F.softmax(self.fc3(dout), dim=1) # 输出层
  return dout
    
    BP(Back Propagation)
    
      
        反向传播（英语：Backpropagation，缩写为BP）指的是计算神经网络参数梯度的方法。总的来说，反向传播依据微积分中的链式法则，沿着从输出层到输入层的顺序，依次计算并存储目标函数有关神经网络各层的中间变量以及参数的梯度。然后用计算出来的梯度和learning rate来更新各层的参数
      
    
  

对输入或输出$X,Y,Z$为任意形状张量的函数$Y=f(X)$和$Z=g(Y)$，通过链式法则，我们有
\(\frac{\partial \mathsf{Z}}{\partial \mathsf{X}} = \text{prod}\left(\frac{\partial \mathsf{Z}}{\partial \mathsf{Y}}, \frac{\partial \mathsf{Y}}{\partial \mathsf{X}}\right),\)
其中prod运算符将根据两个输入的形状，在必要的操作（如转置和互换输入位置）后对两个输入做乘法。

BP代码(Pytorch)
# 计算predict 和 label 之间的loss值
loss = F.mse_loss(pre, ground_truth)  
# 反向传播
loss.backward()  
# 更新权值
w = w - lr * w.grad  
b = b - lr * b.grad

CNN(Convolutional Neural Network)

  
    卷积神经网络（convolutional neural network）是含有卷积层（convolutional layer）的神经网络。
    卷积运算
    卷积层得名于卷积（convolution）运算，在二维卷积层中，一个二维输入数组和一个二维核（kernel）数组通过互相关运算输出一个二维数组。如图所示，输入是一个高和宽均为3的二维数组。我们将该数组的形状记为(3,3), 核数组的高和宽分别为2。该数组在卷积计算中又称卷积核或过滤器(filter)。卷积核窗口（又称卷积窗口）的形状取决于卷积核的高和宽，即(2,2)。图中的阴影部分为第一个输出元素及其计算所使用的输入和核数组元素：$0×0+1×1+3×2+4×3=190×0+1×1+3×2+4×3=19$。

    填充(padding)
    填充（padding）是指在输入高和宽的两侧填充元素（通常是0元素）。图中在原输入高和宽的两侧分别添加了值为0的元素，使得输入高和宽从3变成了5，并导致输出高和宽由2增加到4。图中的阴影部分为第一个输出元素及其计算所使用的输入和核数组元素:0×0+0×1+0×2+0×3=00×0+0×1+0×2+0×3=0。
  

步幅(stride)
刚才的例子里，在高和宽两个方向上步幅均为1。我们也可以使用更大步幅。下图展示了在高上步幅为3、在宽上步幅为2的二维互相关运算。可以看到，输出第一列第二个元素时，卷积窗口向下滑动了3行，而在输出第一行第二个元素时卷积窗口向右滑动了2列。当卷积窗口在输入上再向右滑动2列时，由于输入元素无法填满窗口，无结果输出。图中的阴影部分为输出元素及其计算所使用的输入和核数组元素：$0×0+0×1+1×2+2×3=80×0+0×1+1×2+2×3=8、0×0+6×1+0×2+0×3=60×0+6×1+0×2+0×3=6$。

池化(pooling)
池化层每次对输入数据的一个固定形状窗口（又称池化窗口）中的元素计算输出。不同于卷积层里计算输入和核的互相关性，池化层直接计算池化窗口内元素的最大值或者平均值。该运算也分别叫做最大池化或平均池化。在二维最大池化中，池化窗口从输入数组的最左上方开始，按从左往右、从上往下的顺序，依次在输入数组上滑动。当池化窗口滑动到某一位置时，窗口中的输入子数组的最大值即输出数组中相应位置的元素。

这里把取最大换成取平均的话就可以达到平均池化的效果
CNN代码(Pytorch)
这里以处理图像来举例
# LeNet卷积神经网络代码
class LeNet(nn.Module):
    def __init__(self):
        super(LeNet, self).__init__()
        self.conv = nn.Sequential(
            nn.Conv2d(1, 6, 5), # in_channels, out_channels, kernel_size
            nn.Sigmoid(),
            nn.MaxPool2d(2, 2), # kernel_size, stride
            nn.Conv2d(6, 16, 5),
            nn.Sigmoid(),
            nn.MaxPool2d(2, 2)
        )
        self.fc = nn.Sequential(
            nn.Linear(16*4*4, 120),
            nn.Sigmoid(),
            nn.Linear(120, 84),
            nn.Sigmoid(),
            nn.Linear(84, 10)
        )

    def forward(self, img):
        feature = self.conv(img)
        output = self.fc(feature.view(img.shape[0], -1)) # flatten
        return output

模型框架
LeNet(
  (conv): Sequential(
    (0): Conv2d(1, 6, kernel_size=(5, 5), stride=(1, 1))
    (1): Sigmoid()
    (2): MaxPool2d(kernel_size=2, stride=2, padding=0, dilation=1, ceil_mode=False)
    (3): Conv2d(6, 16, kernel_size=(5, 5), stride=(1, 1))
    (4): Sigmoid()
    (5): MaxPool2d(kernel_size=2, stride=2, padding=0, dilation=1, ceil_mode=False)
  )
  (fc): Sequential(
    (0): Linear(in_features=256, out_features=120, bias=True)
    (1): Sigmoid()
    (2): Linear(in_features=120, out_features=84, bias=True)
    (3): Sigmoid()
    (4): Linear(in_features=84, out_features=10, bias=True)
  )
)

LSTM(Long Short-Term Memory)

  
    LSTM 中引入了3个门，即输入门（input gate）、遗忘门（forget gate）和输出门（output gate），以及与隐藏状态形状相同的记忆细胞（某些文献把记忆细胞当成一种特殊的隐藏状态），从而记录额外的信息。
  

如图所示，长短期记忆的门的输入均为当前时间步输入$X_t$与上一时间步隐藏状态$H_{t−1}$，输出由激活函数为sigmoid函数的全连接层计算得到。如此一来，这3个门元素的值域均为[0,1]。长短期记忆需要计算候选记忆细胞$\tilde{C_t}$。它的计算与上面介绍的3个门类似，但使用了值域在[−1,1][−1,1]的tanh函数作为激活函数。候选记忆细胞和输入门结合后再与上一个时间点的记忆细胞和遗忘门结合的结果加和得到当前时间点的$C_t$

\[\boldsymbol{C}_t = \boldsymbol{F}t \odot \boldsymbol{C}{t-1} + \boldsymbol{I}_t \odot \tilde{\boldsymbol{C}}_t.\]

有了记忆细胞$C_t$以后，接下来还可以通过输出门来控制从记忆细胞到隐藏状态$\boldsymbol{H}_t$的信息的流动：

\(\boldsymbol{H}_t = \boldsymbol{O}_t \odot \text{tanh}(\boldsymbol{C}_t).\)

  符号⊙表示元素乘法
    LSTM代码(Pytorch)
    ```python
x = torch.rand(t,b,c)  # seq_len, batch_size, input_size(135, 42, 39)
lstm = nn.LSTM(input_size = 39, hidden_size = 128, num_layers = 2)
out, hidden = lstm(x)
  


out.shape:torch.Size([135, 42, 128])  # 每一个时间步骤的最后一层隐藏层
hidden[0].shape:torch.Size([2, 42, 128])  # 最后一个时间步骤的每一层的隐藏单元
hidden[1].shape: torch.Size([2, 42, 128])  # 最后一个时间步骤的每一层的隐藏单元所对应的memory cell
### LSTM代码(Source)
```python
class LSTMCell(torch.nn.Module):

    def __init__(self, input_size, hidden_size):
        super().__init__()
        self.input_size = input_size
        self.hidden_size = hidden_size

        self.ft = torch.nn.Linear(input_size+hidden_size, hidden_size)
        self.it = torch.nn.Linear(input_size+hidden_size, hidden_size)
        self.zt = torch.nn.Linear(input_size+hidden_size, hidden_size)
        self.ot = torch.nn.Linear(input_size+hidden_size, hidden_size)

    def forward(self, x, state):
        '''
        :param x: [B, input_size]
        :return: h[B, hidden_size], c[B, hidden_size]
        '''
        h, c = state  # h(t-1), c(t-1)
        input = torch.cat([x,h],dim=1) # [B, input_size+hidden_size]

        ft = self.ft(input).sigmoid()
        it = self.it(input).sigmoid()
        zt = self.zt(input).tanh()
        ot = self.ot(input).sigmoid()

        ct = ft * c + it * zt
        ht = ot * ct.tanh()
        return ht, ct

class LSTM(torch.nn.Module):

    def __init__(self, input_size, hidden_size):
        super().__init__()
        self.cell = LSTMCell(input_size, hidden_size)

    def forward(self, x, state=None):
        '''
        :param x: [B, T, C]
        '''
        if state is None:
            h = torch.zeros(x.shape[0], self.cell.hidden_size,
                            device=x.device)
            c = torch.zeros(x.shape[0], self.cell.hidden_size,
                            device=x.device)
        else:
            h, c = state

        inputs = x.unbind(1)
        outputs = []
        for i in range(len(inputs)):
            h, c = self.cell(inputs[i], (h, c))
            outputs.append(h)
        return torch.stack(outputs, dim=1), (h,c)

x = torch.randn(5, 3, 768)
y = x.unbind(1)
for temp in y:
    print(temp.shape)
输出:      
torch.Size([5, 768])
torch.Size([5, 768])
torch.Size([5, 768])

def lstm(inputs, state, params):
    [W_xi, W_hi, b_i, W_xf, W_hf, b_f, W_xo, W_ho, b_o, W_xc, W_hc, b_c, W_hq, b_q] = params
    (H, C) = state
    outputs = []
    for X in inputs:
        I = torch.sigmoid(torch.matmul(X, W_xi) + torch.matmul(H, W_hi) + b_i)
        F = torch.sigmoid(torch.matmul(X, W_xf) + torch.matmul(H, W_hf) + b_f)
        O = torch.sigmoid(torch.matmul(X, W_xo) + torch.matmul(H, W_ho) + b_o)
        C_tilda = torch.tanh(torch.matmul(X, W_xc) + torch.matmul(H, W_hc) + b_c)
        C = F * C + I * C_tilda
        H = O * C.tanh()
        Y = torch.matmul(H, W_hq) + b_q
        outputs.append(Y)
    return outputs, (H, C)


  这两种是两种不同的结合方式去得到gate，一种是加和的方式，一种是concat的方式


GRU(Gated Recurrent Unit)

  
    它引入了重置门（reset gate）和更新门（update gate）的概念
  




门控循环单元中的重置门和更新门的输入均为当前时间步输入$X_t$
 与上一时间步隐藏状态$H_{t−1}$ ,输出由激活函数为sigmoid函数的全连接层计算得到。门控循环单元将计算候选隐藏状态来辅助稍后的隐藏状态计算。将当前时间步重置门的输出与上一时间步隐藏状态做按元素乘法(符号为$\odot$)。然后，将按元素乘法的结果与当前时间步的输入连结，再通过含激活函数tanh的全连接层计算出候选隐藏状态，其所有元素的值域为$[-1, 1]$。
 最后，时间步$t$的隐藏状态$\boldsymbol{H}_t \in \mathbb{R}^{n \times h}$的计算使用当前时间步的更新门$\boldsymbol{Z}t$来对上一时间步的隐藏状态$\boldsymbol{H}{t-1}$和当前时间步的候选隐藏状态$\tilde{\boldsymbol{H}}_t$做组合：

\[\boldsymbol{H}_t = \boldsymbol{Z}t \odot \boldsymbol{H}{t-1} + (1 - \boldsymbol{Z}_t) \odot \tilde{\boldsymbol{H}}_t.\]

GRU代码(Pytorch)
x = torch.rand(t,b,c)  # seq_len, batch_size, input_size(135, 42, 39)
gru = nn.GRU(input_size = 39, hidden_size = 128, num_layers = 2)
out, hidden = gru(x)

out.shape:torch.Size([135, 42, 128])  # 每一个时间步骤的最后一层隐藏层
hidden.shape:torch.Size([2, 42, 128])  # 最后一个时间步骤的每一层的隐藏单元

GRU代码(Source)
def gru(inputs, state, params):
    W_xz, W_hz, b_z, W_xr, W_hr, b_r, W_xh, W_hh, b_h, W_hq, b_q = params
    H, = state
    outputs = []
    for X in inputs:
        Z = torch.sigmoid(torch.matmul(X, W_xz) + torch.matmul(H, W_hz) + b_z)
        R = torch.sigmoid(torch.matmul(X, W_xr) + torch.matmul(H, W_hr) + b_r)
        H_tilda = torch.tanh(torch.matmul(X, W_xh) + torch.matmul(R * H, W_hh) + b_h)
        H = Z * H + (1 - Z) * H_tilda
        Y = torch.matmul(H, W_hq) + b_q
        outputs.append(Y)
    return outputs, (H,)

ResNet(残差网络)[Optional]

  
    假设我们希望学出的理想映射为$f(x)$，下图左图虚线框中的部分需要直接拟合出该映射$f(x)$，而右图虚线框中的部分则需要拟合出有关恒等映射的残差映射$f(x)−x$。残差映射在实际中往往更容易优化。

残差网络由多个残差块(Block)组合而成的。残差块里首先有2个有相同输出通道数的3×3卷积层。每个卷积层后接一个批量归一化层和ReLU激活函数。然后我们将输入跳过这两个卷积运算后直接加在最后的ReLU激活函数前。这样的设计要求两个卷积层的输出与输入形状一样，从而可以相加。如果想改变通道数，就需要引入一个额外的1×1卷积层来将输入变换成需要的形状后再做相加运算。(长度保持一致，通道数变回一致)
    Residual Block 代码(Pytorch)
    ```python
class Residual(nn.Module): 
    def init(self, in_channels, out_channels, use_1x1conv=False, stride=1):
        super(Residual, self).init()
        self.conv1 = nn.Conv2d(in_channels, out_channels, kernel_size=3, padding=1, stride=stride)
        self.conv2 = nn.Conv2d(out_channels, out_channels, kernel_size=3, padding=1)
        if use_1x1conv:
            self.conv3 = nn.Conv2d(in_channels, out_channels, kernel_size=1, stride=stride)
        else:
            self.conv3 = None
        self.bn1 = nn.BatchNorm2d(out_channels)
        self.bn2 = nn.BatchNorm2d(out_channels)
  


def forward(self, X):
    Y = F.relu(self.bn1(self.conv1(X)))
    Y = self.bn2(self.conv2(Y))
    if self.conv3:
        X = self.conv3(X)
    return F.relu(Y + X) ``` * block堆叠起来就形成了Resnet网络 ## GAN(Generative Adversarial Network)[Optional] >> 模型主要由两个模块构成，分别为判别器(discriminator)与生成器(generator)，如下图所示。其中，判别器判断一个样本是真实样本还是生成器生成样本；生成器生成样本，并尽量让判别器无法判断是否是生成的。




算法过程是这样的，首先Generator和Discriminator都随机初始化参数，然后先fix住Generator的参数，然后随机sample一个random noise给Generator, 让Generator生成一个图片，生成的图片为假图片(label -1), 然后再拿同数量的ground truth图片(label 1)去训练Discriminator，Discriminator训练好以后，fix住Discriminator的参数，这时候再让Generator生成的图片在Disciminator中的值越接近1越好为目标来更新Generator的参数。
GAN代码(Pytorch)
class Generator(nn.Module):
